### ISSUE TITLE
Make `From ... Require ...` behave like -Q rather than -R

### ISSUE THREAD

**Issue by gmalecha** [coq/coq#11124]
#### Description of the problem

I commonly have directory structures that look like this (maybe this is frowned upon?):

```sh
- Internal
|  |- foo.v
|  |- bar.v
|- foo.v
|- bar.v
```

I would like clients to be able to do:

```coq
From X Require Import foo bar.
```

But this often ends up importing `Internal.foo` and `Internal.bar`, which is quite annoying.

I imagine changing this could be problematic from a backwards compatibility point of view, so I'm opening this issue for comments. I'm happy to implement the change if there is a possibility for it to get merged.

#### Coq Version

```sh
The Coq Proof Assistant, version 8.9.1 (June 2019)
compiled on Jun 4 2019 22:11:06 with OCaml 4.07.1
```

**Comment by Zimmi48** [comment]
I think @ejgallego is interested in adding support for internal / private sub-libraries.

I think your title is very misleading however, since `-Q` vs `-R` is rather about the `From` part.

**Comment by silene** [comment]
Those are typical lines at the start of each of my `.v` files:

    From Coq Require Import Reals Bool ZArith Psatz List.
    From Flocq Require Import Core Digits Bracket Round Operations.

I would be severely annoyed if I had to type instead:

    From Coq Require Import Reals.Reals Bool.Bool ZArith.ZArith micromega.Psatz Lists.List.
    From Flocq Require Import Core.Core Core.Digits Calc.Bracket Calc.Round Calc.Operations.

So, it is not just a matter of backward compatibility. It is also a matter of plain usability.

**Comment by silene** [comment]
That said, you could fix your issue by tweaking the search algorithm so that it favors shorter paths, which I would be perfectly fine with.

**Comment by gmalecha** [comment]
@silene It seems like this is a product of how libraries are written. If the Coq standard library defined `Coq.List` as re-exporting the public parts of the `Coq.Lists.List` module, then everything would work out fine. Libraries like `Lists.List`, `Bool.Bool`, `ZArith.ZArith` seem really awkward to me since you are literally repeating the same piece of information twice. As for `Flocq`, `Core.Core` seems to have the same property.

Favoring shorter paths would probably be a somewhat workable solution, but there are still issues. For example, if `Internal.foo` has already been built but `foo` has not, then `coq` will pull the wrong one. There are also potentially annoying pieces if files are not added to your `_CoqProject`.

**Comment by silene** [comment]
I agree. This is definitely a product of how libraries are written. If you did not give the same name to two different files of your library, the issue would not have occurred in the first place. Different naming conventions, different issues.

**Comment by ppedrot** [comment]
This implicit naming feature is important I think, as explained by @silene. We could provide a syntax to force the name to be absolute though. Currently, `From X Require Y.` stands for the regex `X.*.Y`, it is only a matter of also allowing to specify that `Y` should be absolute.

(Also, as already noticed this has nothing to do with -Q / -R, so I'm changing the title.)

**Comment by ejgallego** [comment]
Indeed we could make the `From` syntax more flexible so you would have to do `From foo.*` to require recursively, as done in other languages.

**Comment by gmalecha** [comment]
I think @ejgallego 's proposal would be reasonable, it does still change the default behavior though so that might be a non-starter for @silene. My feeling is that there essentially needs to be harmony in the way these things work together. A few small-ish tweaks to libraries would make the absolute import thing work reasonably well but if those changes are off-limits then we need to consider other things.

The reason that I don't like recursive searching is that adding a file in one project somewhere can completely break clients, it doesn't respect weakening. Globally unique names are a bad solution to this, Coq (thankfully) doesn't insist on globally unique names. I think that fuzzy searches are similarly problematic.

**Comment by ejgallego** [comment]
IMVHO a mixture of what the Java/Typescript/Python world do for name resolution would be a good thing to adopt for Coq; these folks have plenty of experience with quite large developments so their approach has been tested. I agree with @gmalecha in that using `from foo.*` is quite fragile, so it think that moving Coq's default towards the non-recursive variant is well motivated.

We should indeed go thru a slow deprecation phase, so `--strict-paths` does activate the more strict behavior, and at some point we warn the users, then at some point we switch `--strict-paths` to be the default.

We should also take that opportunity to remove the implicit qualification of the `Coq` prefix.

**Comment by JasonGross** [comment]
Is this a duplicate of #9839?

**Comment by herbelin** [comment]
Would @SkySkimmer's proposal on coq-club `Require .X.Y` be an easily implementable satisfactory backwards-compatible answer (or an alternative syntax to it if ever someone has better to propose). I'd be ready to bet that even developers would be happy that we find a concrete answer to this issue.

**Comment by RalfJung** [comment]
The syntax seems reasonable to me, assuming it fits the existing grammar.

**Comment by silene** [comment]
Do you intend to allow the `.X` syntax for any qualified name? Or just for `Require`? Because we also have open reports about the general case. It would be great that, whichever syntax is chosen for `Require`, it can be later used for any qualified name. Otherwise, we will have yet another syntax inconsistency we have to live with.

**Comment by herbelin** [comment]
> Do you intend to allow the `.X` syntax for any qualified name? Or just for `Require`? Because we also have open reports about the general case. It would be great that, whichever syntax is chosen for `Require`, it can be later used for any qualified name. Otherwise, we will have yet another syntax inconsistency we have to live with.

Yes, I thlnk it makes sense (and fits the existing grammar). It seems that we just have to mentally get accustomized to this syntactic "innovation".

**Comment by SkySkimmer** [comment]
One concern with an absolute path syntax is that it would break the minimizer. Typically suppose you have a file with logical name `Lib.A` containing `Definition v := 0.`, and a `bug.v` file to minimize containing
~~~coq
Require Lib.A.

Check eq_refl : Lib.A.v = 1.
~~~
the minimizer will want to inline A turning the file into something like
~~~coq
(* I don't remember the exact mangling the minimizer uses but it's something like this *)
Module Export __FILE_Lib_A.
Module Export Lib.
Module A.
Definition v := 0.
End A.
End Lib.
End __FILE_Lib_A.

Check eq_refl : Lib.A.v = 1.
~~~
which works (ie preserves the error), but now if the bug used the absolute syntax `Check eq_refl : .Lib.A.v = 1.` instead it won't work.

A similar trick is used for shadowing names for compat, eg https://github.com/rocq-prover/rocq/blob/f4e25061958980d0fc28144cd3b40587ba6c7724/theories/Ltac2/Compat/Coq818.v#L7-L12 which wouldn't work if a user referred to `.Ltac2.Array.empty` instead of `Ltac2.Array.empty`.

cc @JasonGross @andres-erbsen

**Comment by JasonGross** [comment]
As long as absolute paths are only permitted for variants of `Require`, the minimizer will not have any issues.  Allowing it in general will indeed cause issues.

**Comment by ejgallego** [comment]
Another possibility would be to improve the minimizer, for example using LSP / Flèche we could do a lot more stuff.

For example, it would be convenient if the minimizer could work over set of files, removing inlining thus.

**Comment by SkySkimmer** [comment]
It's nice to get a single file at the end though.
